File: nodejs/src/bordas/carro-borda.js
const MsgErro = require('../entidades/msg-erro');
const Carro = require('../entidades/db/carro');
const { Op } = require('sequelize');

function init(app) {
  const base = '/carro';
  app.post(base+'', carro_borda_save);
  app.get(base+'/todos', carro_borda_get_todos);
  app.get(base+'', carro_borda_get_um);
}
init.init = init;
init.carro_borda_save = carro_borda_save;
init.carro_borda_get_todos = carro_borda_get_todos;
init.carro_borda_get_um = carro_borda_get_um;

module.exports = init;

// -------------------------------------------------------

async function carro_borda_save(req, res) {
  if( !req.body.nome ) return res.status(400).send(new MsgErro(0,"Campo 'nome' (string) obrigatório!"));
  if( !req.body.cor ) return res.status(400).send(new MsgErro(0,"Campo 'cor' (string) obrigatório!"));
  if( !req.body.ano ) return res.status(400).send(new MsgErro(0,"Campo 'ano' (int) obrigatório!"));
  const carro = await Carro.create({ 
    nome: req.body.nome,
    cor: req.body.cor,
    ano: req.body.ano,
    pessoa_id: req.body.dono || null
  });
  return res.status(200).json(carro);
}

async function carro_borda_get_todos(req, res) {
  const page = parseInt(req.query.page) || 0;
  const size = parseInt(req.query.size) || 10;
  const lista = await Carro.findAll({ include: 'dono', offset: page*size, limit: size });
  return res.status(200).json(lista);
}

async function carro_borda_get_um(req, res) {
  const page = parseInt(req.query.page) || 0;
  const size = parseInt(req.query.size) || 10;
  const id = parseInt(req.query.id) || 0;
  const nome = req.query.nome || '';
  const cor = req.query.cor || '';
  const ano = parseInt(req.query.ano) || 0;
  const where = {};
  if( id ) where.id = id;
  if( nome.trim() ) where.nome = { [Op.like]: `%${nome.trim()}%` };
  if( cor.trim() ) where.cor = { [Op.like]: `%${cor.trim()}%` };
  if( ano ) where.ano = ano;
  const lista = await Carro.findAll({ include: 'dono', offset: page*size, limit: size, where:where });
  return res.status(200).json(lista);
}

File: golang/src/bordas/carro-borda.go
package bordas

import (
	"docker-memory/golang/src/config"
	"docker-memory/golang/src/entidades"
	"docker-memory/golang/src/entidades/db"
	"log"
	"strings"

	"github.com/gin-gonic/gin"
)

func InitCarro(r *gin.Engine) {
	base := "/carro"
	r.POST(base+"/", carroBorda_post)
	r.GET(base+"/todos", carroBorda_get_todos)
	r.GET(base+"/", carroBorda_get)
}

// ----------------------------------------------------------

func carroBorda_post(c *gin.Context) {
	var params db.Carro
	c.ShouldBindJSON(&params)
	if params.Nome == "" {
		c.JSON(400, entidades.MsgErro{Msg: "Campo 'nome' (string) obrigatório!"})
		return
	}
	if params.Cor == "" {
		c.JSON(400, entidades.MsgErro{Msg: "Campo 'cor' (string) obrigatório!"})
		return
	}
	if params.Ano == 0 {
		c.JSON(400, entidades.MsgErro{Msg: "Campo 'ano' (string) obrigatório!"})
		return
	}

	res := config.DB.Create(&params)
	if res.Error != nil {
		log.Println(res.Error.Error())
		c.JSON(500, entidades.MsgErro{Msg: "Erro ao guardar dados!"})
		return
	}

	c.JSON(200, params)
}

func carroBorda_get_todos(c *gin.Context) {
	var params entidades.Paging
	c.ShouldBindQuery(&params)
	params.CheckDefaults()

	var carros []db.Carro
	config.DB.Limit(params.Size).Offset(params.Page * params.Size).Joins("Dono").Find(&carros)
	c.JSON(200, carros)
}

func carroBorda_get(c *gin.Context) {
	var params entidades.Paging
	c.ShouldBindQuery(&params)
	params.CheckDefaults()
	var carroParams db.Carro
	c.ShouldBindQuery(&carroParams)

	sql := config.DB.Limit(params.Size).Offset(params.Page * params.Size).Joins("Dono")

	if carroParams.Id > 0 {
		sql = sql.Where("carro.id = ?", carroParams.Id)
	}
	carroParams.Nome = strings.TrimSpace(carroParams.Nome)
	if carroParams.Nome != "" {
		sql = sql.Where("carro.nome LIKE ?", "%"+carroParams.Nome+"%")
	}
	carroParams.Cor = strings.TrimSpace(carroParams.Cor)
	if carroParams.Cor != "" {
		sql = sql.Where("cor LIKE ?", "%"+carroParams.Cor+"%")
	}
	if carroParams.Ano > 0 {
		sql = sql.Where("ano = ?", carroParams.Ano)
	}

	var carros []db.Carro
	sql.Find(&carros)
	c.JSON(200, carros)
}
